apps/api/v2/src/modules/slots/slots-2024-09-04/services/slots.service.ts

```diff
@@ -215,6 +223,59 @@ export class SlotsService_2024_09_04 {
     }
   }
 
+  async validateRoundRobinSlotAvailability(
+    eventTypeId: number,
+    startDate: DateTime,
+    endDate: DateTime,
+    hosts: Host[]
+  ) {
+    const fixedHosts = hosts.filter((host) => host.isFixed === true);
+    const nonFixedHosts = hosts.filter((host) => host.isFixed === false);
+
+    if (fixedHosts.length > 0) {
+      await this.validateFixedHostsAvailability(eventTypeId, startDate, endDate, fixedHosts);
+    } else {
+      await this.validateNonFixedHostsAvailability(eventTypeId, startDate, endDate, nonFixedHosts);
+    }
+  }
+
+  async validateFixedHostsAvailability(
+    eventTypeId: number,
+    startDate: DateTime,
+    endDate: DateTime,
+    hosts: Host[]
+  ) {
+    const existingBooking = await this.slotsRepository.getExistingBooking(
+      eventTypeId,
+      startDate.toJSDate(),
+      endDate.toJSDate()
+    );
+    const hasHostAsAttendee = hosts.some((host) =>
+      existingBooking?.attendees.some((attendee) => attendee.id === host.userId)
+    );
+
+    if (hasHostAsAttendee) {
+      throw new UnprocessableEntityException(`Can't reserve a slot if the event is already booked.`);
+    }
+  }
+
+  async validateNonFixedHostsAvailability(
+    eventTypeId: number,
+    startDate: DateTime,
+    endDate: DateTime,
+    hosts: Host[]
+  ) {
+    const existingSlotReservations = await this.slotsRepository.getExistingSlotsReservationCount(
+      eventTypeId,
+      startDate.toISO() ?? "",
+      endDate.toISO() ?? ""
+    );
+
+    if (existingSlotReservations === hosts.length) {
+      throw new UnprocessableEntityException(`Can't reserve a slot since the team has no available hosts.`);
```

Comment:
Can we change the error message to focusing on "event type" not that the team has no available hosts:
```
Can't reserve the slot because the event type has no available hosts left at this time slot.
```